# Priprema podataka

[Izvorni bilježnik preuzet iz *Data Science: Introduction to Machine Learning for Data Science Python and Machine Learning Studio by Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## Istraživanje informacija o `DataFrame`

> **Cilj učenja:** Na kraju ovog pododjeljka trebali biste se osjećati ugodno u pronalaženju općih informacija o podacima pohranjenim u pandas DataFrameovima.

Kada učitate svoje podatke u pandas, vrlo je vjerojatno da će biti u obliku `DataFrame`. Međutim, ako vaš `DataFrame` sadrži 60.000 redaka i 400 stupaca, kako uopće početi dobivati osjećaj za ono s čime radite? Srećom, pandas pruža nekoliko praktičnih alata za brzo pregledavanje općih informacija o `DataFrameu`, uz mogućnost pregleda prvih nekoliko i posljednjih nekoliko redaka.

Kako bismo istražili ovu funkcionalnost, uvest ćemo Python biblioteku scikit-learn i koristiti jedan poznati skup podataka koji je svaki data scientist vidio stotine puta: skup podataka britanskog biologa Ronalda Fishera *Iris* korišten u njegovom radu iz 1936. godine "The use of multiple measurements in taxonomic problems":


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
Učitali smo Iris Dataset u varijablu `iris_df`. Prije nego što se detaljnije pozabavimo podacima, bilo bi korisno znati broj podatkovnih točaka koje imamo i ukupnu veličinu skupa podataka. Korisno je pogledati obujam podataka s kojima radimo.


In [2]:
iris_df.shape

(150, 4)

Dakle, imamo 150 redaka i 4 stupca podataka. Svaki redak predstavlja jednu podatkovnu točku, a svaki stupac predstavlja jednu značajku povezanu s podatkovnim okvirom. Dakle, u osnovi, postoji 150 podatkovnih točaka koje sadrže po 4 značajke.

`shape` je ovdje atribut podatkovnog okvira, a ne funkcija, zbog čega ne završava s parom zagrada.


### `DataFrame.columns`
Sada ćemo se posvetiti 4 stupca podataka. Što točno svaki od njih predstavlja? Atribut `columns` će nam dati nazive stupaca u dataframeu.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

Kao što možemo vidjeti, postoje četiri(4) stupca. Atribut `columns` nam govori imena stupaca i zapravo ništa drugo. Ovaj atribut postaje važan kada želimo identificirati značajke koje skup podataka sadrži.


### `DataFrame.info`
Količina podataka (dana atributom `shape`) i nazivi značajki ili stupaca (dani atributom `columns`) daju nam određene informacije o skupu podataka. Sada bismo željeli detaljnije istražiti skup podataka. Funkcija `DataFrame.info()` vrlo je korisna za to.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


Odavde možemo uočiti nekoliko stvari:
1. Tip podataka za svaki stupac: U ovom skupu podataka svi su podaci pohranjeni kao 64-bitni brojevi s pomičnim zarezom.
2. Broj vrijednosti koje nisu null: Rješavanje null vrijednosti važan je korak u pripremi podataka. Time ćemo se pozabaviti kasnije u bilježnici.


### DataFrame.describe()
Recimo da imamo puno numeričkih podataka u našem skupu podataka. Jednostavne statističke kalkulacije poput srednje vrijednosti, medijana, kvartila itd. mogu se izračunati za svaki stupac pojedinačno. Funkcija `DataFrame.describe()` pruža nam statistički sažetak numeričkih stupaca skupa podataka.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

Gornji izlaz prikazuje ukupan broj podatkovnih točaka, srednju vrijednost, standardnu devijaciju, minimum, donji kvartil (25%), medijan (50%), gornji kvartil (75%) i maksimalnu vrijednost svakog stupca.


### `DataFrame.head`
Uz sve gore navedene funkcije i atribute, dobili smo pregled na visokoj razini skupa podataka. Znamo koliko ima podataka, koliko ima značajki, koji je tip podataka svake značajke i broj vrijednosti koje nisu null za svaku značajku.

Sada je vrijeme da pogledamo same podatke. Pogledajmo kako izgledaju prvih nekoliko redaka (prvih nekoliko točaka podataka) našeg `DataFrame`:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

Kao što vidimo u izlazu, prikazano je pet (5) unosa iz skupa podataka. Ako pogledamo indeks s lijeve strane, otkrivamo da su to prvih pet redaka.


### Vježba:

Iz gornjeg primjera jasno je da `DataFrame.head` prema zadanim postavkama vraća prvih pet redaka `DataFramea`. Možete li u donjoj ćeliji koda smisliti način za prikaz više od pet redaka?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
Drugi način pregledavanja podataka može biti od kraja (umjesto od početka). Suprotno od `DataFrame.head` je `DataFrame.tail`, koji vraća zadnjih pet redaka `DataFrame`:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

U praksi je korisno moći lako pregledati prvih nekoliko redaka ili posljednjih nekoliko redaka `DataFrame`-a, osobito kada tražite odstupanja u uređenim skupovima podataka.

Sve funkcije i atributi prikazani gore uz pomoć primjera koda pomažu nam da steknemo uvid u podatke.

> **Zaključak:** Već i samo pregledavanjem metapodataka o informacijama u DataFrame-u ili prvih i posljednjih nekoliko vrijednosti u njemu možete odmah dobiti ideju o veličini, obliku i sadržaju podataka s kojima radite.


### Nedostajući podaci
Pogledajmo pobliže nedostajuće podatke. Nedostajući podaci javljaju se kada neka vrijednost nije pohranjena u nekim stupcima.

Uzmimo primjer: recimo da je netko osjetljiv na svoju težinu i ne ispuni polje za težinu u anketi. Tada će vrijednost težine za tu osobu nedostajati.

U većini slučajeva, u stvarnim skupovima podataka, javljaju se nedostajuće vrijednosti.

**Kako Pandas obrađuje nedostajuće podatke**

Pandas obrađuje nedostajuće vrijednosti na dva načina. Prvi ste već vidjeli u prethodnim odjeljcima: `NaN`, ili Not a Number (nije broj). Ovo je zapravo posebna vrijednost koja je dio IEEE specifikacije za brojeve s pomičnim zarezom i koristi se samo za označavanje nedostajućih vrijednosti s pomičnim zarezom.

Za nedostajuće vrijednosti koje nisu brojevi s pomičnim zarezom, pandas koristi Python objekt `None`. Iako se može činiti zbunjujućim da ćete naići na dvije različite vrste vrijednosti koje u suštini znače isto, postoje opravdani programerski razlozi za ovaj izbor dizajna, a u praksi, ovakav pristup omogućuje pandas biblioteci da pruži dobar kompromis u velikoj većini slučajeva. Unatoč tome, i `None` i `NaN` imaju ograničenja na koja morate obratiti pažnju u vezi s načinom na koji se mogu koristiti.


### `None`: nedostajući podaci koji nisu tipa float
Budući da `None` dolazi iz Pythona, ne može se koristiti u NumPy i pandas nizovima koji nisu tipa podataka `'object'`. Zapamtite, NumPy nizovi (i strukture podataka u pandas) mogu sadržavati samo jednu vrstu podataka. To im daje ogromnu snagu za rad s velikim količinama podataka i računalne operacije, ali također ograničava njihovu fleksibilnost. Takvi nizovi moraju se "promijeniti" u "najniži zajednički nazivnik", odnosno tip podataka koji može obuhvatiti sve u nizu. Kada je `None` prisutan u nizu, to znači da radite s Python objektima.

Da biste to vidjeli u praksi, razmotrite sljedeći primjer niza (obratite pažnju na `dtype`):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

Stvarnost promjene tipova podataka prema višem tipu nosi sa sobom dvije nuspojave. Prvo, operacije će se izvršavati na razini interpretiranog Python koda, a ne kompajliranog NumPy koda. U suštini, to znači da će sve operacije koje uključuju `Series` ili `DataFrame` s vrijednostima `None` biti sporije. Iako vjerojatno nećete primijetiti ovaj pad performansi, kod velikih skupova podataka to bi moglo postati problem.

Druga nuspojava proizlazi iz prve. Budući da `None` u osnovi vraća `Series` ili `DataFrame` u svijet običnog Pythona, korištenje NumPy/pandas agregacija poput `sum()` ili `min()` na nizovima koji sadrže vrijednost ``None`` općenito će rezultirati pogreškom:


In [10]:
example1.sum()

TypeError: ignored

**Ključna točka**: Zbrajanje (i druge operacije) između cijelih brojeva i vrijednosti `None` nije definirano, što može ograničiti rad s skupovima podataka koji ih sadrže.


### `NaN`: nedostajuće vrijednosti tipa float

Za razliku od `None`, NumPy (a samim time i pandas) podržava `NaN` za svoje brze, vektorizirane operacije i ufuncs. Loša vijest je da svaka aritmetička operacija izvedena na `NaN` uvijek rezultira `NaN`. Na primjer:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

Dobre vijesti: agregacije koje se izvode na nizovima s `NaN` u njima ne uzrokuju pogreške. Loše vijesti: rezultati nisu jednako korisni:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### Vježba:


In [11]:
# What happens if you add np.nan and None together?


Zapamtite: `NaN` je samo za nedostajuće vrijednosti s pomičnim zarezom; ne postoji ekvivalent `NaN` za cijele brojeve, stringove ili Booleove vrijednosti.


### `NaN` i `None`: null vrijednosti u pandas

Iako se `NaN` i `None` mogu ponašati malo drugačije, pandas je ipak dizajniran da ih obrađuje naizmjenično. Da biste vidjeli na što mislimo, razmotrite `Series` s cijelim brojevima:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### Vježba:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


U procesu promjene tipova podataka na više razine kako bi se uspostavila homogenost podataka u `Series` i `DataFrame` objektima, pandas će bez problema zamijeniti nedostajuće vrijednosti između `None` i `NaN`. Zbog ove značajke dizajna, korisno je razmišljati o `None` i `NaN` kao dvije različite vrste "null" vrijednosti u pandas. Zapravo, neki od osnovnih metoda koje ćete koristiti za rad s nedostajućim vrijednostima u pandas odražavaju ovu ideju u svojim nazivima:

- `isnull()`: Generira Booleovu masku koja označava nedostajuće vrijednosti
- `notnull()`: Suprotno od `isnull()`
- `dropna()`: Vraća filtriranu verziju podataka
- `fillna()`: Vraća kopiju podataka s popunjenim ili imputiranim nedostajućim vrijednostima

Ovo su važne metode koje treba savladati i s kojima se treba osjećati ugodno, pa ćemo ih detaljnije obraditi.


### Otkrivanje null vrijednosti

Sada kada smo razumjeli važnost nedostajućih vrijednosti, trebamo ih otkriti u našem skupu podataka prije nego što ih obradimo. 
I `isnull()` i `notnull()` su osnovne metode za otkrivanje null podataka. Obje vraćaju Booleanske maske preko vaših podataka.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Pogledajte pažljivo rezultat. Iznenađuje li vas nešto? Iako je `0` aritmetička nula, ipak je potpuno valjan cijeli broj i pandas ga tako tretira. `''` je malo suptilniji. Iako smo ga u Poglavlju 1 koristili za predstavljanje prazne vrijednosti stringa, ipak je objekt tipa string i nije prikaz null vrijednosti prema pandas-u.

Sada, okrenimo perspektivu i koristimo ove metode na način koji je bliži stvarnoj praksi. Boolean maske možete koristiti izravno kao indeks ``Series`` ili ``DataFrame``, što može biti korisno kada radite s izoliranim nedostajućim (ili prisutnim) vrijednostima.

Ako želimo ukupan broj nedostajućih vrijednosti, jednostavno možemo napraviti zbroj maske koju generira metoda `isnull()`.


In [19]:
example3.isnull().sum()

2

### Vježba:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**Ključna točka**: I metode `isnull()` i `notnull()` daju slične rezultate kada ih koristite u DataFrameovima: prikazuju rezultate i indeks tih rezultata, što će vam uvelike pomoći dok se bavite svojim podacima.


### Rješavanje nedostajućih podataka

> **Cilj učenja:** Na kraju ovog pododjeljka trebali biste znati kako i kada zamijeniti ili ukloniti null vrijednosti iz DataFrameova.

Modeli strojnog učenja ne mogu sami obraditi nedostajuće podatke. Stoga, prije nego što podatke proslijedimo modelu, moramo se pozabaviti tim nedostajućim vrijednostima.

Način na koji se nedostajući podaci obrađuju nosi sa sobom suptilne kompromise, može utjecati na vašu konačnu analizu i rezultate u stvarnom svijetu.

Postoje dva glavna načina rješavanja nedostajućih podataka:

1.   Uklanjanje retka koji sadrži nedostajuću vrijednost
2.   Zamjena nedostajuće vrijednosti nekom drugom vrijednošću

Razmotrit ćemo obje metode i njihove prednosti i nedostatke detaljno.


### Uklanjanje null vrijednosti

Količina podataka koju prosljeđujemo našem modelu ima izravan utjecaj na njegovu učinkovitost. Uklanjanje null vrijednosti znači da smanjujemo broj podataka, a time i veličinu skupa podataka. Stoga je preporučljivo ukloniti retke s null vrijednostima kada je skup podataka prilično velik.

Drugi slučaj može biti da određeni redak ili stupac ima mnogo nedostajućih vrijednosti. Tada ih je možda najbolje ukloniti jer ne bi značajno pridonijeli našoj analizi s obzirom na to da većina podataka za taj redak/stupac nedostaje.

Osim identificiranja nedostajućih vrijednosti, pandas pruža praktičan način za uklanjanje null vrijednosti iz `Series` i `DataFrame` objekata. Da bismo to vidjeli u praksi, vratimo se na `example3`. Funkcija `DataFrame.dropna()` pomaže u uklanjanju redaka s null vrijednostima.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Imajte na umu da bi ovo trebalo izgledati kao vaš izlaz iz `example3[example3.notnull()]`. Razlika ovdje je u tome što je, umjesto samo indeksiranja na maskirane vrijednosti, `dropna` uklonio te nedostajuće vrijednosti iz `Series` `example3`.

Budući da DataFrames imaju dvije dimenzije, nude više opcija za uklanjanje podataka.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(Jeste li primijetili da je pandas promijenio tip podataka za dva stupca u float kako bi prilagodio `NaN` vrijednosti?)

Ne možete ukloniti pojedinačnu vrijednost iz `DataFrame`, već morate ukloniti cijele redove ili stupce. Ovisno o tome što radite, možda ćete htjeti učiniti jedno ili drugo, a pandas vam nudi opcije za oba. Budući da u znanosti o podacima stupci obično predstavljaju varijable, a redovi predstavljaju opažanja, češće ćete uklanjati redove podataka; zadana postavka za `dropna()` je uklanjanje svih redova koji sadrže bilo kakve null vrijednosti:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

Ako je potrebno, možete ukloniti NA vrijednosti iz stupaca. Koristite `axis=1` za to:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

Primijetite da ovo može ukloniti puno podataka koje možda želite zadržati, posebno u manjim skupovima podataka. Što ako želite ukloniti samo retke ili stupce koji sadrže nekoliko ili čak sve null vrijednosti? Te postavke možete odrediti u `dropna` pomoću parametara `how` i `thresh`.

Prema zadanim postavkama, `how='any'` (ako želite sami provjeriti ili vidjeti koje druge parametre metoda ima, pokrenite `example4.dropna?` u ćeliji koda). Alternativno, možete odrediti `how='all'` kako biste uklonili samo retke ili stupce koji sadrže sve null vrijednosti. Proširimo naš primjer `DataFrame` kako bismo to vidjeli u praksi u sljedećoj vježbi.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> Ključne točke:  
1. Uklanjanje null vrijednosti dobra je ideja samo ako je skup podataka dovoljno velik.  
2. Cijeli redovi ili stupci mogu se ukloniti ako im nedostaje većina podataka.  
3. Metoda `DataFrame.dropna(axis=)` pomaže u uklanjanju null vrijednosti. Argument `axis` označava hoće li se uklanjati redovi ili stupci.  
4. Može se koristiti i argument `how`. Po zadanim postavkama postavljen je na `any`. Dakle, uklanja samo one redove/stupce koji sadrže bilo kakve null vrijednosti. Može se postaviti na `all` kako bi se odredilo da ćemo ukloniti samo one redove/stupce gdje su sve vrijednosti null.  


### Vježba:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


Parametar `thresh` omogućuje precizniju kontrolu: postavljate broj *ne-null* vrijednosti koje redak ili stupac mora imati kako bi bio zadržan:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

Ovdje su prva i zadnja redak uklonjeni jer sadrže samo dvije vrijednosti koje nisu null.


### Popunjavanje null vrijednosti

Ponekad ima smisla popuniti nedostajuće vrijednosti onima koje bi mogle biti valjane. Postoji nekoliko tehnika za popunjavanje null vrijednosti. Prva je korištenje znanja o domeni (znanja o temi na kojoj se temelji skup podataka) kako bi se na neki način približno odredile nedostajuće vrijednosti.

Možete koristiti `isnull` za popunjavanje vrijednosti izravno, ali to može biti zamorno, osobito ako imate puno vrijednosti za popuniti. Budući da je ovo tako čest zadatak u znanosti o podacima, pandas pruža `fillna`, koji vraća kopiju `Series` ili `DataFrame` s nedostajućim vrijednostima zamijenjenim vrijednostima po vašem izboru. Napravimo još jedan primjer `Series` kako bismo vidjeli kako ovo funkcionira u praksi.


### Kategorijski podaci (ne-numerički)
Prvo ćemo razmotriti ne-numeričke podatke. U skupovima podataka imamo stupce s kategorijskim podacima, npr. Spol, Istina ili Laž itd.

U većini ovih slučajeva, nedostajuće vrijednosti zamjenjujemo `modom` stupca. Recimo, imamo 100 podataka, od kojih je 90 označilo Istina, 8 označilo Laž, a 2 nisu popunjena. Tada možemo popuniti ta 2 s Istina, uzimajući u obzir cijeli stupac.

Opet, ovdje možemo koristiti znanje iz domene. Razmotrimo primjer popunjavanja pomoću moda.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

Sad, prvo pronađimo modus prije nego što popunimo vrijednost `None` s modusom.


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

Dakle, zamijenit ćemo None s True


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

Kao što možemo vidjeti, null vrijednost je zamijenjena. Nepotrebno je reći, mogli smo napisati bilo što umjesto `'True'` i to bi bilo zamijenjeno.


### Numerički podaci
Sada prelazimo na numeričke podatke. Ovdje imamo dva uobičajena načina za zamjenu nedostajućih vrijednosti:

1. Zamjena s medijanom reda
2. Zamjena s prosjekom reda

Medijan koristimo u slučaju da su podaci iskrivljeni i sadrže ekstremne vrijednosti. To je zato što je medijan otporan na ekstremne vrijednosti.

Kada su podaci normalizirani, možemo koristiti prosjek, jer su u tom slučaju prosjek i medijan prilično blizu.

Prvo, uzmimo stupac koji je normalno distribuiran i popunimo nedostajuće vrijednosti prosjekom tog stupca.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

Srednja vrijednost stupca je


In [33]:
np.mean(fill_with_mean[0])

0.0

Popunjavanje s prosjekom


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

Kao što možemo vidjeti, nedostajuća vrijednost je zamijenjena njenim prosjekom.


Sad pokušajmo s drugim dataframeom, i ovaj put ćemo zamijeniti vrijednosti None s medijanom stupca.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

Medijan drugog stupca je


In [36]:
fill_with_median[1].median()

4.0

Popunjavanje s medijanom


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

Kao što možemo vidjeti, NaN vrijednost je zamijenjena medianom stupca.


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

Možete popuniti sve prazne unose jednom vrijednošću, poput `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> Ključne točke:
1. Popunjavanje nedostajućih vrijednosti treba provesti kada postoji malo podataka ili kada postoji strategija za popunjavanje nedostajućih podataka.
2. Znanje o domeni može se koristiti za popunjavanje nedostajućih vrijednosti njihovim približavanjem.
3. Kod kategorijskih podataka, nedostajuće vrijednosti najčešće se zamjenjuju modom stupca.
4. Kod numeričkih podataka, nedostajuće vrijednosti obično se popunjavaju srednjom vrijednošću (za normalizirane skupove podataka) ili medijanom stupca.


### Vježba:


In [40]:
# What happens if you try to fill null values with a string, like ''?


Možete **popuniti unaprijed** null vrijednosti, što znači koristiti posljednju valjanu vrijednost za popunjavanje null vrijednosti:


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

Također možete **popuniti unatrag** kako biste propagirali sljedeću valjanu vrijednost unatrag za popunjavanje null vrijednosti:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

Kao što možete pretpostaviti, ovo funkcionira isto s DataFrameovima, ali možete također odrediti `axis` duž kojeg ćete popuniti null vrijednosti:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

Primijetite da kada prethodna vrijednost nije dostupna za popunjavanje unaprijed, null vrijednost ostaje.


### Vježba:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


Možete biti kreativni u načinu na koji koristite `fillna`. Na primjer, pogledajmo ponovno `example4`, ali ovaj put popunimo nedostajuće vrijednosti prosjekom svih vrijednosti u `DataFrame`:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

Primijetite da stupac 3 još uvijek nema vrijednosti: zadani smjer je popunjavanje vrijednosti po redovima.

> **Zaključak:** Postoji više načina za rješavanje nedostajućih vrijednosti u vašim skupovima podataka. Specifična strategija koju koristite (uklanjanje, zamjena ili način zamjene) trebala bi biti određena specifičnostima tih podataka. Razvijat ćete bolji osjećaj za rješavanje nedostajućih vrijednosti što više radite i komunicirate s podacima.


### Kodiranje kategorijskih podataka

Modeli strojnog učenja rade isključivo s brojevima i bilo kojim oblikom numeričkih podataka. Ne mogu razlikovati "Da" od "Ne", ali mogu razlikovati 0 od 1. Dakle, nakon popunjavanja nedostajućih vrijednosti, potrebno je kodirati kategorijske podatke u neki numerički oblik kako bi ih model mogao razumjeti.

Kodiranje se može obaviti na dva načina. O njima ćemo raspravljati u nastavku.


**KODIRANJE OZNAKA**

Kodiranje oznaka u osnovi znači pretvaranje svake kategorije u broj. Na primjer, recimo da imamo skup podataka o putnicima zrakoplovnih kompanija i postoji stupac koji sadrži njihovu klasu među sljedećim ['poslovna klasa', 'ekonomska klasa', 'prva klasa']. Ako se provede kodiranje oznaka, to bi se transformiralo u [0,1,2]. Pogledajmo primjer putem koda. Budući da ćemo učiti `scikit-learn` u nadolazećim bilježnicama, ovdje ga nećemo koristiti.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Da bismo izvršili kodiranje oznaka na prvom stupcu, prvo moramo opisati mapiranje svake klase na broj, prije zamjene


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

Kao što možemo vidjeti, izlaz odgovara onome što smo očekivali. Dakle, kada koristimo kodiranje oznaka? Kodiranje oznaka koristi se u jednom ili oba sljedeća slučaja:
1. Kada je broj kategorija velik
2. Kada su kategorije poredane.


**JEDNOSTRUKO KODIRANJE (ONE HOT ENCODING)**

Druga vrsta kodiranja je Jednostruko Kodiranje (One Hot Encoding). Kod ove vrste kodiranja, svaka kategorija iz stupca dodaje se kao zaseban stupac, a svaka podatkovna točka dobiva vrijednost 0 ili 1 ovisno o tome sadrži li tu kategoriju. Dakle, ako postoji n različitih kategorija, n stupaca će biti dodano u podatkovni okvir.

Na primjer, uzmimo isti primjer klasa u avionu. Kategorije su bile: ['business class', 'economy class', 'first class']. Dakle, ako primijenimo jednostruko kodiranje, sljedeća tri stupca će biti dodana u skup podataka: ['class_business class', 'class_economy class', 'class_first class'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Hajdemo izvršiti one hot encoding na prvom stupcu


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

Svaki one hot kodirani stupac sadrži 0 ili 1, što označava postoji li ta kategorija za tu podatkovnu točku.


Kada koristimo one hot encoding? One hot encoding se koristi u jednom ili oba sljedeća slučaja:

1. Kada je broj kategorija i veličina skupa podataka manja.
2. Kada kategorije ne slijede nikakav određeni redoslijed.


> Ključne točke:
1. Kodiranje se koristi za pretvaranje nenumeričkih podataka u numeričke podatke.
2. Postoje dvije vrste kodiranja: Label kodiranje i One Hot kodiranje, a obje se mogu primijeniti ovisno o zahtjevima skupa podataka.


## Uklanjanje dupliciranih podataka

> **Cilj učenja:** Na kraju ovog pododjeljka trebali biste se osjećati ugodno u prepoznavanju i uklanjanju dupliciranih vrijednosti iz DataFrameova.

Osim nedostajućih podataka, često ćete naići na duplicirane podatke u stvarnim skupovima podataka. Srećom, pandas pruža jednostavan način za otkrivanje i uklanjanje dupliciranih unosa.


### Identificiranje duplikata: `duplicated`

Duplikate možete lako uočiti pomoću metode `duplicated` u pandas, koja vraća Booleovu masku koja pokazuje je li unos u `DataFrame` duplikat ranijeg unosa. Napravimo još jedan primjer `DataFrame` kako bismo vidjeli kako to funkcionira.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### Uklanjanje duplikata: `drop_duplicates`
`drop_duplicates` jednostavno vraća kopiju podataka za koje su sve vrijednosti označene kao `duplicated` postavljene na `False`:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

Oba `duplicated` i `drop_duplicates` prema zadanim postavkama uzimaju u obzir sve stupce, ali možete odrediti da pregledavaju samo podskup stupaca u vašem `DataFrame`:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **Zaključak:** Uklanjanje dupliciranih podataka ključni je dio gotovo svakog projekta u području podatkovne znanosti. Duplicirani podaci mogu promijeniti rezultate vaših analiza i dati netočne rezultate!


## Provjere kvalitete podataka u stvarnom svijetu

> **Cilj učenja:** Na kraju ovog dijela trebali biste biti sposobni prepoznati i ispraviti uobičajene probleme kvalitete podataka u stvarnom svijetu, uključujući nekonzistentne kategorijske vrijednosti, nenormalne numeričke vrijednosti (iznimke) i duplicirane entitete s varijacijama.

Iako su nedostajuće vrijednosti i točne duplicirane vrijednosti uobičajeni problemi, skupovi podataka iz stvarnog svijeta često sadrže suptilnije probleme:

1. **Nekonzistentne kategorijske vrijednosti**: Ista kategorija napisana na različite načine (npr. "USA", "U.S.A", "United States")
2. **Nenormalne numeričke vrijednosti**: Ekstremne iznimke koje ukazuju na pogreške pri unosu podataka (npr. dob = 999)
3. **Gotovo duplicirani redovi**: Zapisi koji predstavljaju isti entitet s malim varijacijama

Istražimo tehnike za otkrivanje i rješavanje ovih problema.


### Stvaranje uzorka "neurednog" skupa podataka

Prvo, stvorimo uzorak skupa podataka koji sadrži vrste problema s kojima se često susrećemo u stvarnim podacima:


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. Otkrivanje nedosljednih kategorijskih vrijednosti

Primijetite da stupac `country` ima više prikaza za iste zemlje. Idemo identificirati te nedosljednosti:


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### Standardizacija kategorijskih vrijednosti

Možemo stvoriti mapiranje za standardizaciju ovih vrijednosti. Jednostavan pristup je pretvoriti u mala slova i stvoriti rječnik za mapiranje:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**Alternativa: Korištenje nepreciznog podudaranja**

Za složenije slučajeve, možemo koristiti neprecizno podudaranje nizova s bibliotekom `rapidfuzz` kako bismo automatski otkrili slične nizove:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. Otkrivanje neuobičajenih numeričkih vrijednosti (iznimki)

Promatrajući stupac `age`, imamo neke sumnjive vrijednosti poput 199 i -5. Koristit ćemo statističke metode za otkrivanje ovih iznimki.


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### Korištenje metode IQR (interkvartilnog raspona)

Metoda IQR je pouzdana statistička tehnika za otkrivanje odstupanja koja je manje osjetljiva na ekstremne vrijednosti:


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### Korištenje Z-score metode

Metoda Z-score identificira odstupanja na temelju standardnih devijacija od srednje vrijednosti:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### Rukovanje odstupanjima

Nakon što su otkrivena, odstupanja se mogu riješiti na nekoliko načina:
1. **Ukloniti**: Izbaciti retke s odstupanjima (ako su pogreške)
2. **Ograničiti**: Zamijeniti graničnim vrijednostima
3. **Zamijeniti s NaN**: Tretirati kao nedostajuće podatke i koristiti tehnike imputacije
4. **Zadržati**: Ako su legitimne ekstremne vrijednosti


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. Otkrivanje gotovo identičnih redaka

Primijetite da naš skup podataka ima više unosa za "John Smith" s malo različitim vrijednostima. Identificirajmo potencijalne duplicirane unose na temelju sličnosti imena.


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### Pronalaženje gotovo identičnih podudarnosti pomoću nejasnog podudaranja

Za naprednije otkrivanje dupliciranih podataka, možemo koristiti nejasno podudaranje za pronalaženje sličnih imena:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### Rukovanje dupliciranim podacima

Nakon što ih identificirate, trebate odlučiti kako postupiti s dupliciranim podacima:
1. **Zadržati prvu pojavu**: Koristite `drop_duplicates(keep='first')`
2. **Zadržati zadnju pojavu**: Koristite `drop_duplicates(keep='last')`
3. **Agregirati informacije**: Kombinirajte informacije iz dupliciranih redaka
4. **Ručno pregledati**: Označite za pregled od strane čovjeka


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### Sažetak: Kompletan proces čišćenja podataka

Složimo sve zajedno u cjelovit proces čišćenja podataka:


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 Vježba izazova

Sad je vaš red! Ispod je novi red podataka s više problema s kvalitetom. Možete li:

1. Identificirati sve probleme u ovom redu
2. Napisati kod za ispravljanje svakog problema
3. Dodati očišćeni red u skup podataka

Evo problematičnih podataka:


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### Ključne točke

1. **Nekonzistentne kategorije** česte su u stvarnim podacima. Uvijek provjerite jedinstvene vrijednosti i standardizirajte ih pomoću mapiranja ili nejasnog podudaranja.

2. **Izvanredne vrijednosti** mogu značajno utjecati na vašu analizu. Koristite znanje o domeni u kombinaciji sa statističkim metodama (IQR, Z-score) za njihovo otkrivanje.

3. **Gotovo duplicirani podaci** teže se otkrivaju od točnih duplikata. Razmislite o korištenju nejasnog podudaranja i normalizacije podataka (pretvaranje u mala slova, uklanjanje razmaka) za njihovo prepoznavanje.

4. **Čišćenje podataka je iterativno**. Možda ćete morati primijeniti više tehnika i pregledati rezultate prije nego što finalizirate očišćeni skup podataka.

5. **Dokumentirajte svoje odluke**. Vodite evidenciju o koracima čišćenja koje ste primijenili i razlozima za to, jer je to važno za reprodukciju i transparentnost.

> **Najbolja praksa:** Uvijek sačuvajte kopiju svojih originalnih "nečistih" podataka. Nikada nemojte prepisivati izvorne datoteke s podacima - kreirajte očišćene verzije s jasnim konvencijama imenovanja, poput `data_cleaned.csv`.



---

**Odricanje od odgovornosti**:  
Ovaj dokument je preveden pomoću AI usluge za prevođenje [Co-op Translator](https://github.com/Azure/co-op-translator). Iako nastojimo osigurati točnost, imajte na umu da automatski prijevodi mogu sadržavati pogreške ili netočnosti. Izvorni dokument na izvornom jeziku treba smatrati autoritativnim izvorom. Za ključne informacije preporučuje se profesionalni prijevod od strane čovjeka. Ne preuzimamo odgovornost za nesporazume ili pogrešna tumačenja koja mogu proizaći iz korištenja ovog prijevoda.
